File: apps/android/app/src/test/java/com/exponential/app/domain/StartedRunMatchTest.kt
package com.exponential.app.domain

import com.exponential.app.data.db.CodingSessionEntity
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

// EXP-536: every start surface navigates into the run it just launched, so
// the row the desktop inserts has to be recognizable from the client. These
// are the rules the six Android surfaces share.
class StartedRunMatchTest {

    // 2026-07-17T12:00:00Z; the watch cuts off at now - SKEW_MS.
    private val nowMs = 1_784_289_600_000L
    private val cutoffMs = nowMs - StartedRunMatch.SKEW_MS

    private fun session(
        id: String = "sess-1",
        issueId: String? = null,
        actionName: String? = null,
        userId: String = "user-1",
        startedAt: String = "2026-07-17T11:59:30Z",
        startedReason: String? = null,
        resumedFromId: String? = null,
    ) = CodingSessionEntity(
        id = id,
        issueId = issueId,
        teamId = "team-1",
        userId = userId,
        actionName = actionName,
        startedReason = startedReason,
        resumedFromId = resumedFromId,
        startedAt = startedAt,
        createdAt = startedAt,
        updatedAt = startedAt,
    )

    private fun matches(session: CodingSessionEntity, key: StartedRunKey) =
        StartedRunMatch.matches(session, key, "user-1", cutoffMs)

    @Test
    fun oneIssueIsASingleRunTwoOrMoreIsABatch() {
        assertNull(StartedRunKey.forIssues(emptyList()))
        assertEquals(StartedRunKey.Issue("a"), StartedRunKey.forIssues(listOf("a")))
        assertEquals(StartedRunKey.Batch, StartedRunKey.forIssues(listOf("a", "b")))
    }

    @Test
    fun singleRunMatchesItsOwnIssueRow() {
        assertTrue(matches(session(issueId = "issue-1"), StartedRunKey.Issue("issue-1")))
        assertFalse(matches(session(issueId = "issue-2"), StartedRunKey.Issue("issue-1")))
    }

    @Test
    fun batchMatchesTheIssuelessActionlessRow() {
        assertTrue(matches(session(), StartedRunKey.Batch))
        // A single-issue run is not the batch we started…
        assertFalse(matches(session(issueId = "issue-1"), StartedRunKey.Batch))
        // …and neither is an action run, which also carries no issue.
        assertFalse(matches(session(actionName = "Fix merge conflicts"), StartedRunKey.Batch))
    }

    @Test
    fun actionRunMatchesOnTheNameSnapshot() {
        val key = StartedRunKey.Action("Fix merge conflicts")
        // action_id is NULL for the builtins, so the name is the only key.
        assertTrue(matches(session(actionName = "Fix merge conflicts"), key))
        assertFalse(matches(session(actionName = "Create action"), key))
    }

    @Test
    fun anIssueRunIgnoresAnActionRunOnTheSameIssue() {
        val row = session(issueId = "issue-1", actionName = "Fix merge conflicts")
        assertFalse(matches(row, StartedRunKey.Issue("issue-1")))
    }

    @Test
    fun teammatesRunsAndOldRowsNeverMatch() {
        val key = StartedRunKey.Issue("issue-1")
        assertFalse(matches(session(issueId = "issue-1", userId = "user-2"), key))
        // Started an hour ago — a pre-existing session, not the one we sent.
        assertFalse(matches(session(issueId = "issue-1", startedAt = "2026-07-17T11:00:00Z"), key))
    }

    /** EXP-530: automation-started rows are the device's own doing — a user's
     * pending start watch must never grab one, even when the action name and
     * timing line up. */
    @Test
    fun automationStartedRowsNeverMatch() {
        val key = StartedRunKey.Action("Fix merge conflicts")
        assertFalse(
            matches(
                session(actionName = "Fix merge conflicts", startedReason = "schedule"),
                key,
            ),
        )
        assertFalse(
            matches(
                session(actionName = "Fix merge conflicts", startedReason = "event"),
                key,
            ),
        )
        // Batch/issue watches reject them too — the reason cut runs first.
        assertFalse(matches(session(startedReason = "schedule"), StartedRunKey.Batch))
        assertFalse(
            matches(
                session(issueId = "issue-1", startedReason = "event"),
                StartedRunKey.Issue("issue-1"),
            ),
        )
    }

    @Test
    fun unparseableStartedAtNeverMatches() {
        // Fail CLOSED here (unlike liveness): navigating into the wrong
        // session is worse than not navigating at all.
        assertFalse(matches(session(issueId = "issue-1", startedAt = "nonsense"), StartedRunKey.Issue("issue-1")))
    }

    // ── EXP-637: a resumed run ──────────────────────────────────────────────

    /** The desktop stamps the ENDED run it continues on the new row, so the
     * match is exact — no name or timing guessing. */
    @Test
    fun resumedRunMatchesOnTheRunItContinues() {
        val key = StartedRunKey.Resumed("sess-old")
        assertTrue(matches(session(resumedFromId = "sess-old"), key))
        assertFalse(matches(session(resumedFromId = "sess-other"), key))
        // A fresh run of the same action is not the resume we sent.
        assertFalse(matches(session(actionName = "Refresh screenshots"), key))
    }

    @Test
    fun resumedRunKeepsTheOwnershipTimingAndAutomationCuts() {
        val key = StartedRunKey.Resumed("sess-old")
        assertFalse(matches(session(resumedFromId = "sess-old", userId = "user-2"), key))
        assertFalse(
            matches(
                session(resumedFromId = "sess-old", startedAt = "2026-07-17T11:00:00Z"),
                key,
            ),
        )
        // An automation that happened to resume something is the device's own
        // doing — the reason cut runs before the key.
        assertFalse(
            matches(
                session(resumedFromId = "sess-old", startedReason = "schedule"),
                key,
            ),
        )
    }
}

File: apps/web/src/lib/started-run-match.test.ts
import { describe, expect, it } from "vitest"
import {
  findStartedRun,
  matchesStartedRun,
  startedRunKeyForIssues,
  type StartedRunCandidate,
  type StartedRunKey,
} from "@/lib/started-run-match"

// EXP-536: every start surface opens the run it just launched, so the row the
// desktop inserts has to be recognizable from the client. The natives mirror
// these rules (`domain/StartedRunMatch.kt`, `Domain/StartedRunMatch.swift`).

const NOW = Date.parse(`2026-07-17T12:00:00Z`)
const CUTOFF = NOW - 120_000

type Row = StartedRunCandidate & { id: string }

function session(over: Partial<Row> = {}): Row {
  return {
    id: `sess-1`,
    issueId: null,
    actionName: null,
    userId: `user-1`,
    startedAt: new Date(NOW - 30_000).toISOString(),
    ...over,
  }
}

const match = (row: Row, key: StartedRunKey) =>
  matchesStartedRun(row, key, `user-1`, CUTOFF)

describe(`startedRunKeyForIssues`, () => {
  it(`is a single run for one id and a batch for two or more`, () => {
    expect(startedRunKeyForIssues([])).toBeNull()
    expect(startedRunKeyForIssues([`a`])).toEqual({ kind: `issue`, issueId: `a` })
    expect(startedRunKeyForIssues([`a`, `b`])).toEqual({ kind: `batch` })
  })
})

describe(`matchesStartedRun`, () => {
  it(`matches a single run on its own issue row`, () => {
    expect(match(session({ issueId: `issue-1` }), { kind: `issue`, issueId: `issue-1` })).toBe(true)
    expect(match(session({ issueId: `issue-2` }), { kind: `issue`, issueId: `issue-1` })).toBe(false)
  })

  it(`matches a batch on the issueless, actionless row`, () => {
    expect(match(session(), { kind: `batch` })).toBe(true)
    // A single-issue run is not the batch we started…
    expect(match(session({ issueId: `issue-1` }), { kind: `batch` })).toBe(false)
    // …and neither is an action run, which also carries no issue.
    expect(match(session({ actionName: `Fix merge conflicts` }), { kind: `batch` })).toBe(false)
  })

  it(`matches an action run on the name snapshot, never the id`, () => {
    const key = { kind: `action`, actionName: `Fix merge conflicts` } as const
    expect(match(session({ actionName: `Fix merge conflicts` }), key)).toBe(true)
    expect(match(session({ actionName: `Create action` }), key)).toBe(false)
  })

  it(`does not mistake an action run on the issue for the issue run`, () => {
    expect(
      match(session({ issueId: `issue-1`, actionName: `Fix merge conflicts` }), {
        kind: `issue`,
        issueId: `issue-1`,
      })
    ).toBe(false)
  })

  it(`ignores teammates' runs and rows older than the skew window`, () => {
    const key = { kind: `issue`, issueId: `issue-1` } as const
    expect(match(session({ issueId: `issue-1`, userId: `user-2` }), key)).toBe(false)
    expect(
      match(
        session({ issueId: `issue-1`, startedAt: new Date(NOW - 3_600_000).toISOString() }),
        key
      )
    ).toBe(false)
  })

  it(`fails closed on an unparseable startedAt`, () => {
    expect(match(session({ issueId: `issue-1`, startedAt: `nonsense` }), {
      kind: `issue`,
      issueId: `issue-1`,
    })).toBe(false)
  })
})

describe(`findStartedRun`, () => {
  it(`picks the matching row out of the collection`, () => {
    const rows = [
      session({ id: `other`, issueId: `issue-9` }),
      session({ id: `mine`, issueId: `issue-1` }),
    ]
    expect(
      findStartedRun(rows, { kind: `issue`, issueId: `issue-1` }, `user-1`, CUTOFF)?.id
    ).toBe(`mine`)
    expect(
      findStartedRun(rows, { kind: `batch` }, `user-1`, CUTOFF)
    ).toBeUndefined()
  })
})
